Add Check Support for Java 21 Record Pattern : HiddenField

child of #14890 
no update is needed for this Issue. we need a new input file and a new UT for record pattern syntax
```
PS D:\CS\test> cat src/Test.java                                                 
public class InputHiddenFieldRecordPattern {
    private String s;
    private int x;
    static int z;

    record Color(String name, int value) { }
    record Point(Color c, int x, int y) { }

    public boolean doStuff(Object f) {
            return f instanceof Color(String s, int x);  // 2 violations
    }


    public void test(Object f) {
        if (f instanceof Color(String s, int x)) { // 2 violations
            System.out.println(s);
        }
    }

   public void test2(Object f) {
        if (f instanceof Point(Color(String s, int x), _, int z)) { // 3 violations
            System.out.println(s);
        }
        switch (f) {
            case Point(Color(String s, int x), _, _) : break; // 2 violations
            default:
                throw new IllegalStateException();
        }
   }
}
PS D:\CS\test> cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
 <module name="TreeWalker">
    <module name="HiddenField">

    </module>
 </module>
</module>
PS D:\CS\test> java  -jar checkstyle-10.14.2-all.jar -c config.xml  src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:10:46: 's' hides a field. [HiddenField]
[ERROR] D:\CS\test\src\Test.java:10:53: 'x' hides a field. [HiddenField]
[ERROR] D:\CS\test\src\Test.java:15:39: 's' hides a field. [HiddenField]
[ERROR] D:\CS\test\src\Test.java:15:46: 'x' hides a field. [HiddenField]
[ERROR] D:\CS\test\src\Test.java:21:45: 's' hides a field. [HiddenField]
[ERROR] D:\CS\test\src\Test.java:21:52: 'x' hides a field. [HiddenField]
[ERROR] D:\CS\test\src\Test.java:21:63: 'z' hides a field. [HiddenField]
[ERROR] D:\CS\test\src\Test.java:25:37: 's' hides a field. [HiddenField]
[ERROR] D:\CS\test\src\Test.java:25:44: 'x' hides a field. [HiddenField]
Audit done.
Checkstyle ends with 9 errors.
PS D:\CS\test> 

```

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/HiddenFieldCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/HiddenFieldCheckTest.java
@@ -494,6 +494,24 @@ public class HiddenFieldCheckTest
                 expected);
     }
 
+    @Test
+    public void testHiddenFieldRecordPattern() throws Exception {
+
+        final String[] expected = {
+            "18:46: " + getCheckMessage(MSG_KEY, "s"),
+            "18:53: " + getCheckMessage(MSG_KEY, "x"),
+            "23:39: " + getCheckMessage(MSG_KEY, "s"),
+            "23:46: " + getCheckMessage(MSG_KEY, "x"),
+            "29:45: " + getCheckMessage(MSG_KEY, "s"),
+            "29:52: " + getCheckMessage(MSG_KEY, "x"),
+            "29:63: " + getCheckMessage(MSG_KEY, "z"),
+            "33:37: " + getCheckMessage(MSG_KEY, "s"),
+            "33:44: " + getCheckMessage(MSG_KEY, "x"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputHiddenFieldRecordPattern.java"), expected);
+    }
+
     /**
      * We cannot reproduce situation when visitToken is called and leaveToken is not.
      * So, we have to use reflection to be sure that even in such situation

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/hiddenfield/InputHiddenFieldRecordPattern.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/hiddenfield/InputHiddenFieldRecordPattern.java
@@ -1,0 +1,38 @@
+/*
+HiddenField
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.coding.hiddenfield;
+
+public class InputHiddenFieldRecordPattern {
+    private String s;
+    private int x;
+    static int z;
+
+    record Color(String name, int value) { }
+    record Point(Color c, int x, int y) { }
+
+    public boolean doStuff(Object f) {
+            return f instanceof Color(String s, int x);  // 2 violations
+    }
+
+
+    public void test(Object f) {
+        if (f instanceof Color(String s, int x)) { // 2 violations
+            System.out.println(s);
+        }
+    }
+
+   public void test2(Object f) {
+        if (f instanceof Point(Color(String s, int x), _, int z)) { // 3 violations
+            System.out.println(s);
+        }
+        switch (f) {
+            case Point(Color(String s, int x), _, _) : break; // 2 violations
+            default:
+                throw new IllegalStateException();
+        }
+   }
+}

